zerver/decorator.py

```diff
@@ -884,6 +887,29 @@ def rate_limit_user(request: HttpRequest, user: UserProfile, domain: str) -> Non
     RateLimitedUser(user, domain=domain).rate_limit_request(request)
 
 
+class TorDataSession(OutgoingSession):
+    def __init__(self) -> None:
+        super().__init__(role="tor_data", timeout=5)
+
+
+@cache_with_key(lambda: "tor_ip_addresses:", timeout=3600)
+def get_tor_ips() -> Set[str]:
+    # We cache this for 1h, based on https://lists.torproject.org/pipermail/tor-project/2020-March/002759.html
+    response = TorDataSession().get("https://check.torproject.org/exit-addresses")
```

Comment:
What's going to happen if this server has an issue causing each request to time out? Given that I think exceptions don't get cached, and we're calling `rate_limit_request_by_ip` on every unauthed request, won't that potentially cause an outage, with every request getting stuck until this times out?

Reply:
Yeah, I think that's a good point; we probably want the fetching to me managed by a cron job.